### PR TITLE
Make it build using OpenSSL 1.1.0

### PR DESCRIPTION
**Pull request by kroeckx**

### CODE REVIEW

**Comment by Diapolo** [comment]
With this patch applied I was able to build ´´libevent-release-2.1.6-beta while linking against ´´openssl-1.1.0b!

File: test/regress_ssl.c [hunk 1]
```diff
@@ -65,6 +65,35 @@
...
+#define BIO_meth_set_destroy(b, f) b->destroy = f
+
+#define BIO_set_init(b, val) b->init = val
+#define BIO_set_data(b, val) b->ptr = val
+#define BIO_set_shutdown(b, val) b->shutdown = val
+#define BIO_get_init(b) b->init
+#define BIO_get_data(b) b->ptr
+#define BIO_get_shutdown(b) b->shutdown
+
+#endif /* OPENSSL_VERSION_NUMBER < 0x10100000L */
```
Comment (azat) [inline test/regress_ssl.c]: The only thing that bother me for now is this copy-paste, can you please move this macros into `openssl-internal.h` (or something like that)?

File: sample/https-client.c [hunk 2]
```diff
@@ -474,6 +474,7 @@ main(int argc, char **argv)
 		SSL_CTX_free(ssl_ctx);
 	if (type == HTTP && ssl)
 		SSL_free(ssl);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
```
Comment (azat) [inline sample/https-client.c]: So seems that initialization also deprecated, can you move add `ifdef` for them too?

**Comment by kroeckx** [comment]
I've pushed a new commit.

**Comment by azat** [comment]
@kroeckx applied, thanks!

After a little rebase:
- tabs instead of whitespaces
- make openssl-compat.h safe for complex expressions
- do not call sk_SSL_COMP_free() in 1.1 (fixes double free)

**Comment by sandrotosi** [comment]
hello @azat, any plan to release a new stable version of libevent with this patch applied?

**Comment by azat** [comment]
Yep, -rc preparation in progress.

**Comment by azat** [comment]
@sandrotosi 2.1.7-rc is out

**Comment by austinnichols101** [comment]
@azat - is there any chance of this being included as a backport for release-2.0.23-beta?

**Comment by azat** [comment]
I would not recommend using 2.0.x, is there anything particular why you are still on 2.0.x?

P.S. it is not that tricky to add openssl 1.1 support for 2.0, but I didn't plan this, since 2.0.x has been out too many years ago

**Comment by austinnichols101** [comment]
It's not critical, just nice to have.  One of our devs is migrating a legacy php application (5.6) that uses libevent 2.0.x and we would like to run it on a modern openssl version.  The app is going away, just not fast enough - if you know what I mean...

**Comment by azat** [comment]
Ok, it was simple enough, the patch had been backported.
Although I'm not sure about any new releases for 2.0.x, can you compile from sources?

**Pull request closed without merging**